design/baremetal-operator/image-builder-integration.md

```diff
@@ -0,0 +1,252 @@
+involves DHCP supplying an IP address and default route.
+
+Some users would prefer not to run DHCP in their infrastructure, but rather use
+static IP addresses instead. In some cases, particularly when provisioning
+using virtualmedia and not using PXE, they might prefer that the network on
+which to contact Ironic is not an untagged VLAN. In many cases, servers will be
+connected to switches that use static link aggregators, and thus should set up
+NIC bonding to enable reliable two-way communication.
+
+Currently there is no way to use static IPs or configure any of the other
+options on a per-host basis because all hosts are booted using the same ramdisk
+image.
+
+### Goals
+
+- Allow each host to use its own deploy ramdisk image customised with an
+  appropriate network configuration.
+
+### Non-Goals
+
+- Allow users who expect to provision a host to specify their own custom image
+  containing IPA.
+- Use Ironic with images that don't contain IPA.
+- Implement custom provisioners not based on Ironic.
+
+## Proposal
+
+### User Stories
+
+As a data center operator, I would like to manage hosts with metal³ and deploy
+them via virtualmedia on networks where there is no DHCP-like configuration
+service.
+
+## Design Details
+
+A new Custom Resource type, `RamDiskImageRequest` will be created in the
+metal3.io group. The Spec will contain the following (optional) fields:
+
+- Name of networkData Secret
+- Architecture
+
+The Status will contain the following fields:
+
+- Image URL
+- Image format (ISO or initrd)
+- Name + Version of networkData Secret
+- Error message
+- Ready condition
+- Error condition
+
+A default controller for this resource type will be implemented that simply
+sets a fixed URL (supplied in its own environment or command line) as the
+output. However, provision will be made for interested parties to implement
+alternate controllers to allow image customisation.
+
+A command-line switch (defaulting to off) will enable integration between the
+baremetal-operator and the image request controller.
+
+### Implementation Details/Notes/Constraints
+
+When integration is enabled and the driver is virtualmedia based (i.e. supports
+an ISO ramdisk format), the controller portion of the baremetal-operator will
+update the `RamDiskImageRequest` with the same name and namespace as the
+`BareMetalHost`, creating it if it doesn't exist, to point to the networkData
+Secret referenced in the `BareMetalHost` Spec. The Architecture will be
+obtained from the Host's hardware profile. (Currently we don't ratify the
+choice of hardware profile until after inspection, but we could move this state
+earlier. In any event, we don't support multiple architectures today, so we
+could use a default value in the interim.)
+
+The ironic provisioner will refuse to report registration complete in
+`ValidateMangementAccess()` until the Ready condition is true in the
+corresponding `RamDiskImageRequest` Status. It will then configure the Node to
```

Comment:
It's not clear under which circumstances the RamDiskImageRequest Status Ready condition will become true. I guess it would be useful to clarify this point, at least for the default controller (I'd assume after it verified and set the Image URL)